File: kamacoder105/kamacoder105.cpp
// dfs

#include<iostream>
#include<vector>
#include<list>
using namespace std;

void dfs(const vector<list<int>>& graph, int key, vector<bool>& visited){
    if (visited[key]){
        return;
    }
    visited[key] = true;

    list<int> keys = graph[key];
    for (int key : keys){
        dfs(graph, key, visited);
    }
}

int main(){
    int n, m, s, t;
    cin >> n >> m;

    vector<list<int>> graph(n + 1);
    while (m--)
    {
        cin >> s >> t;
        graph[s].push_back(t);
    }
    vector<bool> visited(n + 1, false);
    dfs(graph, 1, visited);
    for (int i = 1; i <= n; i++){
        if (visited[i] == false){
            cout << -1 << endl;
            return 0;
        }
    }
    
    cout << 1 << endl;
}


// bfs

#include<iostream>
#include<vector>
#include<list>
#include <queue>
using namespace std;


int main(){
    int n, m, s, t;
    cin >> n >> m;

    vector<list<int>> graph(n + 1);
    while (m--)
    {
        cin >> s >> t;
        graph[s].push_back(t);
    }
    vector<bool> visited(n + 1, false);

    visited[1] = true;
    queue<int> que;
    que.push(1);

    while (!que.empty())
    {
        int key = que.front();
        que.pop();
        list<int> keys = graph[key];

        for (int key : keys){
            if (!visited[key]){
                que.push(key);
                visited[key] = true;
            }
        }
    }
    
    
    for (int i = 1; i <= n; i++){
        if (visited[i] == false){
            cout << -1 << endl;
            return 0;
        }
    }
    
    cout << 1 << endl;
}

File: kamacoder105/kamacoder105.py
# dfs

def dfs(graph, key, visited):
    if visited[key]:
        return
    visited[key] = True
    for neighbor in graph[key]:
        dfs(graph, neighbor, visited)

def main():
    import sys
    input = sys.stdin.read
    data = input().split()

    n = int(data[0])
    m = int(data[1])

    graph = [[] for _ in range(n + 1)]

    index = 2
    for _ in range(m):
        s = int(data[index])
        t = int(data[index + 1])
        graph[s].append(t)
        index += 2
    
    visited = [False] * (n + 1)
    
    dfs(graph, 1, visited)

    for i in range(1, n + 1):
        if not visited[i]:
            print(-1)
            return
    
    print(1)

if __name__ == "__main__":
    main()


# bfs
from collections import deque
def bfs(graph, key, visited):
    if visited[key]:
        return
    que = deque()
    que.append(key)
    visited[key] = True
    while que:
        key = que.popleft()
        keys = graph[key]
        for key in keys:
            if not visited[key]:
                que.append(key)
                visited[key] = True
    
def main():
    import sys
    input = sys.stdin.read
    data = input().split()

    n = int(data[0])
    m = int(data[1])

    graph = [[] for _ in range(n + 1)]

    index = 2
    for _ in range(m):
        s = int(data[index])
        t = int(data[index + 1])
        graph[s].append(t)
        index += 2
    
    visited = [False] * (n + 1)
    
    bfs(graph, 1, visited)

    for i in range(1, n + 1):
        if not visited[i]:
            print(-1)
            return
    
    print(1)

if __name__ == "__main__":
    main()